sdk/lib/opentelemetry/sdk/trace/config/trace_config.rb

```diff
@@ -0,0 +1,109 @@
+          attr_reader :max_links_count
+
+          # The global default max number of attributes per {Event}.
+          attr_reader :max_attributes_per_event
+
+          # The global default max number of attributes per {OpenTelemetry::Trace::Link}.
+          attr_reader :max_attributes_per_link
+
+          # Returns a {TraceConfig} with the desired values.
+          #
+          # @return [TraceConfig] with the desired values.
+          # @raise [ArgumentError] if any of the max numbers are not positive.
+          def initialize(sampler: DEFAULT_SAMPLER,
+                         max_attributes_count: DEFAULT_MAX_ATTRIBUTES_COUNT,
+                         max_events_count: DEFAULT_MAX_EVENTS_COUNT,
+                         max_links_count: DEFAULT_MAX_LINKS_COUNT,
+                         max_attributes_per_event: DEFAULT_MAX_ATTRIBUTES_PER_EVENT,
+                         max_attributes_per_link: DEFAULT_MAX_ATTRIBUTES_PER_LINK)
+            raise ArgumentError, 'max_attributes_count must be positive' unless max_attributes_count.positive?
+            raise ArgumentError, 'max_events_count must be positive' unless max_events_count.positive?
+            raise ArgumentError, 'max_links_count must be positive' unless max_links_count.positive?
+            raise ArgumentError, 'max_attributes_per_event must be positive' unless max_attributes_per_event.positive?
+            raise ArgumentError, 'max_attributes_per_link must be positive' unless max_attributes_per_link.positive?
+
+            @sampler = sampler
+            @max_attributes_count = max_attributes_count
+            @max_events_count = max_events_count
+            @max_links_count = max_links_count
+            @max_attributes_per_event = max_attributes_per_event
+            @max_attributes_per_link = max_attributes_per_link
+          end
+
+          # @api private
+          # Removes oldest entries from a Hash whose size exceeds a limit.
+          #
+          # @param [Hash] attrs This is modified in-place.
+          # @param [Symbol, Integer] limit The maximum number of entries
+          #   allowed in attrs. May be expressed as a number or a symbol
+          #   representing one of the limits in this class.
+          def trim_attributes(attrs, limit)
+            return if attrs.nil?
+
+            limit = public_send(limit) if limit.instance_of?(Symbol)
+            excess = attrs.size - limit
+            excess.times { attrs.shift } if excess.positive?
+            nil
+          end
+
+          # @api private
+          # Removes oldest entries from an Array whose size exceeds a limit.
+          #
+          # @param [Array] arr Array of links or events that is modified in-place.
+          # @param [Symbol, Integer] limit The maximum number of entries
+          #   allowed in the array. May be expressed as a number or a symbol
+          #   representing one of the limits in this class.
+          def trim_links_or_events(arr, limit)
```

Comment:
I'm going to walk back my suggestions from Slack. I don't think the metaprogramming is warranted here. It's convenient for `trim_attributes` because we have 3 different attribute limits, but here both the API and the implementation are simpler with separate `trim_links` and `trim_events`.

Along those lines, I might replace the metaprogramming in `trim_attributes` with simple wrapper methods for each case - that's easier to understand and probably more performant.

Reply:
Done